### ISSUE TITLE
type checking for custom event listening

### ISSUE THREAD

**Issue by andylix** [johnsoncodehk/volar#303]
When I define the emits with the `defineEmits<{(event: 'custom-event', param: number): void}>()` syntax, then doing `v-on:custom-event="handler"` on an element that `emit("custom-event")`, the problem is `handler` currently can be anything except `null`, `undefined`, and a function that accepts more parameters (or different parameters). It can be more useful if the type checking for `v-on:custom-event="handler"` behaves exact like in an assignment statement where it will reject values that doesn't match with its signature `(param: number): void`.

**Comment by johnsoncodehk** [comment]
If I understand correctly, we have do this already?

<img width="875" alt="螢幕截圖 2021-07-14 下午4 09 47" src="https://user-images.githubusercontent.com/16279759/125587230-81d02abd-195c-443f-99a3-827dfe1b1ffc.png">
<img width="1057" alt="螢幕截圖 2021-07-14 下午4 10 16" src="https://user-images.githubusercontent.com/16279759/125587240-24558aa9-b994-498b-8e5a-f2a4406edbe7.png">

**Comment by andylix** [comment]
Sorry, let me rephrase it better. Currently, it works for the following situations:

- null
- undefined
- function with more params
- function with different params

BUT, to make this practical, we need it to report error on these following situations too (which aren't currently working):

- a number
- a string
- an object
- an array

If you have `@custom-event="1"` for example, it will not report the error. (which should be)

**Comment by johnsoncodehk** [comment]
This is the correct behavior, because this is expression and vue compile it to `@custom-event="$event => (1)"`.

**Comment by andylix** [comment]
oh yeah, I totally forgot about that. Thanks for the quick replies.
